tensorboard/backend/application_test.py

```diff
@@ -535,6 +573,57 @@ def testNameToPluginMapping(self):
     self.assertEqual('foo', mapping['foo'].plugin_name)
     self.assertEqual('bar', mapping['bar'].plugin_name)
 
+  def testNormalRoute(self):
+    self._test_route('/data/plugin/foo/foo_route', 200)
+
+  def testNormalRouteIsNotWildcard(self):
+    self._test_route('/data/plugin/foo/foo_route/bogus', 404)
+
+  def testMissingRoute(self):
+    self._test_route('/data/plugin/foo/bogus', 404)
+
+  def testEmptyRoute(self):
+    self._test_route('', 404)
+
+  def testSlashlessRoute(self):
+    self._test_route('runaway', 404)
+
+  def testWildcardAcceptedRoute(self):
+    self._test_route('/data/plugin/bar/wildcard/ok', 200)
+
+  def testLongerWildcardRouteTakesPrecedence(self):
+    # This tests that the longer 'special' wildcard takes precedence over
+    # the shorter one.
+    self._test_route('/data/plugin/bar/wildcard/special/blah', 300)
+
+  def testExactRouteTakesPrecedence(self):
+    # This tests that an exact match takes precedence over a wildcard.
+    self._test_route('/data/plugin/bar/wildcard/special/exact', 200)
+
+  def testWildcardRejectedRoute(self):
+    # A plugin may reject a request passed to it via a wildcard route.
+    # Note our test plugin returns 401 in this case, to distinguish this
+    # response from a 404 passed if the route were not found.
+    self._test_route('/data/plugin/bar/wildcard/bogus', 401)
+
+  def testWildcardRouteWithoutSlash(self):
+    # A wildcard route requires a slash before the '*'.
+    # Lacking one, no route is matched.
+    self._test_route('/data/plugin/bar/wildcard', 404)
+
+  def testEmptyWildcardRouteWithSlash(self):
+    # A wildcard route requires a slash before the '*'.  Here we include the
+    # slash, so we might expect the route to match.
+    #
+    # However: Trailing slashes are automatically removed from incoming requests
+    # in _clean_path().  Consequently, this request does not match the wildcard
```

Comment:
Hmm, interesting. Thanks for including this test case. Is this the
behavior that we actually want? It seems surprising that `/foo/*` does
not match `/foo/`: in glob contexts, `*` always matches the empty
string, so as a plugin author I would not expect this behavior.

The clean-path behavior of always identifying `/foo/` and `/foo` as
user-provided paths is certainly nice. I can see a case for making
pattern `/foo/*` match both paths `/foo/` and `/foo`, and then
disallowing plugins from having conflicting exact and wildcard routes
(e.g., separate `/foo/*` and `/foo/` handlers). How do you think that we
should address this?

Reply:
I don't really like the behavior that clean_path has today, FWIW.  Conflating the request that has a trailing slash and the one without one in general is a bit dangerous because it encourages users to treat the URLs interchangeably, but the semantics of relative paths are different across the two URLs at which the content is served.

If we can establish it's safe, I'd be happy to remove that conversion logic or at least change it into an actual redirect so we aren't serving content under both URLs.  Basically, for every `/foo` handler we could create an implicit `/foo/` redirect handler.  We could either then block the plugin having both `/foo` and `/foo/` handlers (to be consistent), or potentially even block all handlers ending in `/`.

Reply:
Oof, I was assuming that we already did have this redirect handler, à la
GitHub Pages, because in that case clean-path helps mitigate exactly the
confusion that you describe. Agreed, then.

Reply:
Yep, makes sense, but can we punt fixing `/foo` vs. `/foo/` more generally to another PR?  If we leave this test in place, then it will fail when the behavior changes, so we won't forget to revisit it.

Reply:
Okay—it’s not optimal, but if this change unblocks Angular/etc. then I’m
willing to defer.

Reply:
OK.  I think it's topically different enough to merit a separate PR in any case.  Tracking at #2502.